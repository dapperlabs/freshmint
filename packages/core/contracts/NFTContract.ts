// @ts-ignore
import * as fcl from '@onflow/fcl';

// @ts-ignore
import * as t from '@onflow/types';

import * as metadata from '../metadata';
import { FreshmintConfig } from '../config';
import { Transaction, TransactionAuthorizer, TransactionResult, TransactionSigners } from '../transactions';
import { Script } from '../scripts';
import { Path } from '../cadence/values';
import { CommonNFTGenerator } from '../generators/CommonNFTGenerator';
import { HTTPMediaInput, IPFSMediaInput, isIPFSMediaInput } from '../metadata/views';

export interface Royalty {
  address: string;
  receiverPath: string;
  cut: string;
  description?: string;
}

export interface CollectionMetadata {
  name: string;
  description: string;
  url: string;
  squareImage: IPFSMediaInput | HTTPMediaInput;
  bannerImage: IPFSMediaInput | HTTPMediaInput;
  socials: { [name: string]: string };
}

export abstract class NFTContract {
  name: string;
  address?: string;

  schema: metadata.Schema;

  owner?: TransactionAuthorizer;
  payer?: TransactionAuthorizer;
  proposer?: TransactionAuthorizer;

  constructor({
    name,
    address,
    schema,
    owner,
    payer,
    proposer,
  }: {
    name: string;
    address?: string;
    schema: metadata.Schema;
    owner?: TransactionAuthorizer;
    payer?: TransactionAuthorizer;
    proposer?: TransactionAuthorizer;
  }) {
    this.name = name;
    this.address = address;

    this.schema = schema;

    this.owner = owner;
    this.payer = payer;
    this.proposer = proposer;
  }

  setOwner(authorizer: TransactionAuthorizer) {
    this.owner = authorizer;
  }

  setPayer(authorizer: TransactionAuthorizer) {
    this.payer = authorizer;
  }

  setProposer(authorizer: TransactionAuthorizer) {
    this.proposer = authorizer;
  }

  setAddress(address: string) {
    this.address = address;
  }

  getAddress(): string {
    if (this.address !== undefined) {
      return this.address;
    }

    throw new MissingContractAddressError(this.name);
  }

  getSigners(): TransactionSigners {
    const owner = this.owner;
    if (!owner) {
      // TODO: improve error message
      throw 'must specify owner';
    }

    const payer = this.payer ?? owner;
    const proposer = this.proposer ?? owner;

    return {
      payer,
      proposer,
      authorizers: [owner],
    };
  }

  getNFT(address: string, id: string): Script<{ id: string }> {
    return new Script(
      ({ imports }: FreshmintConfig) => {
        const script = CommonNFTGenerator.getNFT({
          imports,
          contractName: this.name,
          contractAddress: this.getAddress(),
        });

        return {
          script,
          args: (arg, t) => [arg(address, t.Address), arg(id, t.UInt64)],
          computeLimit: 9999,
        };
      },
      (result) => result,
    );
  }

  destroyNFT(id: string): Transaction<void> {
    return new Transaction(({ imports }: FreshmintConfig) => {
      const script = CommonNFTGenerator.destroyNFT({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
      });

      return {
        script,
        args: [fcl.arg(id, t.UInt64)],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    }, Transaction.VoidResult);
  }

  setupCollection(authorizer: TransactionAuthorizer): Transaction<void> {
    return new Transaction(({ imports }: FreshmintConfig) => {
      const script = CommonNFTGenerator.setupCollection({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
      });

      return {
        script,
        args: [],
        computeLimit: 9999,
        signers: {
          payer: authorizer,
          proposer: authorizer,
          authorizers: [authorizer],
        },
      };
    }, Transaction.VoidResult);
  }

  transferNFT({ id, toAddress }: { id: string; toAddress: string }): Transaction<void> {
    return new Transaction(({ imports }: FreshmintConfig) => {
      const script = CommonNFTGenerator.transferNFT({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
      });

      return {
        script,
        args: [fcl.arg(toAddress, t.Address), fcl.arg(id, t.UInt64)],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    }, Transaction.VoidResult);
  }

  transferNFTs({ ids, toAddress }: { ids: string[]; toAddress: string }): Transaction<void> {
    return new Transaction(({ imports }: FreshmintConfig) => {
      const script = CommonNFTGenerator.transferNFTs({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
      });

      return {
        script,
        args: [fcl.arg(toAddress, t.Address), fcl.arg(ids, t.Array(t.UInt64))],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    }, Transaction.VoidResult);
  }

  transferQueueToQueue({
    fromQueue,
    toQueue,
    count,
  }: {
    fromQueue: string | null;
    toQueue: string | null;
    count: number;
  }): Transaction<void> {
    return new Transaction(({ imports }: FreshmintConfig) => {
      const script = CommonNFTGenerator.transferQueueToQueue({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
      });

      return {
        script,
        args: [
          fcl.arg(fromQueue, t.Optional(t.String)),
          fcl.arg(toQueue, t.Optional(t.String)),
          fcl.arg(count.toString(), t.Int),
        ],
        computeLimit: 9999,
        signers: this.getSigners(),
      };
    }, Transaction.VoidResult);
  }

  transferQueueToCollection({
    fromQueue,
    toAddress,
    count,
  }: {
    fromQueue: string | null;
    toAddress: string;
    count: number;
  }): Transaction<string[]> {
    return new Transaction(
      ({ imports }: FreshmintConfig) => {
        const script = CommonNFTGenerator.transferQueueToCollection({
          imports,
          contractName: this.name,
          contractAddress: this.getAddress(),
        });

        return {
          script,
          args: [
            fcl.arg(fromQueue, t.Optional(t.String)),
            fcl.arg(toAddress, t.Address),
            fcl.arg(count.toString(), t.Int),
          ],
          computeLimit: 9999,
          signers: this.getSigners(),
        };
      },
      ({ events }: TransactionResult) => {
        const deposits = events.filter((event) => event.type.includes('.Deposit'));
        return deposits.map((event) => event.data.id);
      },
    );
  }

  getCollectionMetadata(): Script<CollectionMetadata | null> {
    return new Script(({ imports }: FreshmintConfig) => {
      const script = CommonNFTGenerator.getCollectionMetadata({
        imports,
        contractName: this.name,
        contractAddress: this.getAddress(),
      });

      return {
        script,
        args: () => [],
        computeLimit: 9999,
      };
    }, parseCollectionMetadataResult);
  }

  getRoyalties(): Script<Royalty[]> {
    return new Script(
      ({ imports }: FreshmintConfig) => {
        const script = CommonNFTGenerator.getRoyalties({
          imports,
          contractName: this.name,
          contractAddress: this.getAddress(),
        });

        return {
          script,
          args: () => [],
          computeLimit: 9999,
        };
      },
      (royalties) => {
        return royalties.map((royalty: any) => ({
          address: royalty.receiver.address,
          receiverPath: new Path(royalty.receiver.path.value).toString(),
          // Trim trailing zeros (0.01000 becomes 0.01)
          cut: parseFloat(royalty.cut).toString(),
          description: royalty.description !== '' ? royalty.description : undefined,
        }));
      },
    );
  }
}

export class MissingContractAddressError extends Error {
  contractName: string;

  constructor(contractName: string) {
    const message = `Missing contract address for contract with name "${contractName}".`;

    super(message);

    this.contractName = contractName;
  }
}

export function prepareRoyalties(royalties: Royalty[]): {
  royaltyAddresses: string[];
  royaltyReceiverPaths: Path[];
  royaltyCuts: string[];
  royaltyDescriptions: string[];
} {
  const royaltyAddresses = royalties.map((royalty) => royalty.address);
  const royaltyReceiverPaths = royalties.map((royalty) => Path.fromString(royalty.receiverPath));
  const royaltyCuts = royalties.map((royalty) => royalty.cut);
  const royaltyDescriptions = royalties.map((royalty) => royalty.description ?? '');

  return { royaltyAddresses, royaltyReceiverPaths, royaltyCuts, royaltyDescriptions };
}

function parseCollectionMetadataResult(result: any | null): CollectionMetadata | null {
  if (result === null) {
    return null;
  }

  return {
    name: result.name,
    description: result.description,
    url: result.externalURL.url,
    squareImage: parseMedia(result.squareImage),
    bannerImage: parseMedia(result.bannerImage),
    socials: parseSocials(result.socials),
  };
}

function parseMedia(media: { mediaType: string; file: IPFSFile | HTTPFile }): IPFSMediaInput | HTTPMediaInput {
  if (isIPFSFile(media.file)) {
    return {
      ipfs: media.file.path ? { cid: media.file.cid, path: media.file.path } : media.file.cid,
      type: media.mediaType,
    };
  }

  return {
    url: media.file.url,
    type: media.mediaType,
  };
}

interface IPFSFile {
  cid: string;
  path: string | null;
}

function isIPFSFile(file: IPFSFile | HTTPFile): file is IPFSFile {
  return (file as IPFSFile).cid !== undefined;
}

interface HTTPFile {
  url: string;
}

function parseSocials(socials: { [key: string]: { url: string } }): { [key: string]: string } {
  const result: { [key: string]: string } = {};

  for (const key in socials) {
    result[key] = socials[key].url;
  }

  return result;
}

export function prepareCollectionMetadata(metadataViewsAddress: string, collectionMetadata: CollectionMetadata) {
  return prepareStruct(`A.${fcl.sansPrefix(metadataViewsAddress)}.MetadataViews.NFTCollectionDisplay`, [
    prepareStructField('name', prepareString(collectionMetadata.name)),
    prepareStructField('description', prepareString(collectionMetadata.description)),
    prepareStructField('externalURL', prepareExternalURL(metadataViewsAddress, collectionMetadata.url)),
    prepareStructField('squareImage', prepareMedia(metadataViewsAddress, collectionMetadata.squareImage)),
    prepareStructField('bannerImage', prepareMedia(metadataViewsAddress, collectionMetadata.bannerImage)),
    prepareStructField('socials', prepareSocials(metadataViewsAddress, collectionMetadata.socials)),
  ]);
}

function prepareStruct(typeId: string, fields: any[]) {
  return {
    type: 'Struct',
    value: {
      id: typeId,
      fields,
    },
  };
}

function prepareStructField(name: string, value: any) {
  return {
    name,
    value,
  };
}

function prepareDictionary(pairs: any[]) {
  return {
    type: 'Dictionary',
    value: pairs,
  };
}

function prepareString(value: string) {
  return {
    type: 'String',
    value,
  };
}

function prepareOptional(value: any | null) {
  return {
    type: 'Optional',
    value,
  };
}

function prepareExternalURL(metadataViewsAddress: string, url: string) {
  return prepareStruct(`A.${fcl.sansPrefix(metadataViewsAddress)}.MetadataViews.ExternalURL`, [
    prepareStructField('url', prepareString(url)),
  ]);
}

function prepareMedia(metadataViewsAddress: string, media: IPFSMediaInput | HTTPMediaInput) {
  const file = prepareFile(metadataViewsAddress, media);

  return prepareStruct(`A.${fcl.sansPrefix(metadataViewsAddress)}.MetadataViews.Media`, [
    prepareStructField('file', file),
    prepareStructField('mediaType', prepareString(media.type)),
  ]);
}

function prepareFile(metadataViewsAddress: string, media: IPFSMediaInput | HTTPMediaInput) {
  if (isIPFSMediaInput(media)) {
    return prepareIPFSFile(metadataViewsAddress, media);
  }

  return prepareHTTPFile(metadataViewsAddress, media);
}

function prepareIPFSFile(metadataViewsAddress: string, media: IPFSMediaInput) {
  if (typeof media.ipfs === 'string') {
    return prepareStruct(`A.${fcl.sansPrefix(metadataViewsAddress)}.MetadataViews.IPFSFile`, [
      prepareStructField('cid', prepareString(media.ipfs)),
      prepareStructField('path', prepareOptional(null)),
    ]);
  }

  return prepareStruct(`A.${fcl.sansPrefix(metadataViewsAddress)}.MetadataViews.IPFSFile`, [
    prepareStructField('cid', prepareString(media.ipfs.cid)),
    prepareStructField('path', prepareOptional(media.ipfs.path ? prepareString(media.ipfs.path) : null)),
  ]);
}

function prepareHTTPFile(metadataViewsAddress: string, media: HTTPMediaInput) {
  return prepareStruct(`A.${fcl.sansPrefix(metadataViewsAddress)}.MetadataViews.HTTPFile`, [
    prepareStructField('url', prepareString(media.url)),
  ]);
}

function prepareSocials(metadataViewsAddress: string, socials: { [key: string]: string }) {
  const pairs = [];

  for (const key in socials) {
    pairs.push({ key: prepareString(key), value: prepareExternalURL(metadataViewsAddress, socials[key]) });
  }

  return prepareDictionary(pairs);
}
