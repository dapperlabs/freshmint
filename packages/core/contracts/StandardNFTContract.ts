// @ts-ignore
import * as fcl from '@onflow/fcl';
// @ts-ignore
import * as t from '@onflow/types';

import { NFTContract, CollectionMetadata, prepareCollectionMetadata, Royalty, prepareRoyalties } from './NFTContract';
import { hashMetadata, MetadataMap } from '../metadata';
import { StandardNFTGenerator } from '../generators/StandardNFTGenerator';
import { FreshmintConfig, ContractImports } from '../config';
import { PublicKey, SignatureAlgorithm, HashAlgorithm } from '../crypto';
import { Transaction, TransactionResult } from '../transactions';

export type NFTMintResult = {
  id: string;
  metadata: MetadataMap;
  transactionId: string;
};

export class StandardNFTContract extends NFTContract {
  getSource(imports: ContractImports, options?: { saveAdminResourceToContractAccount?: boolean }): string {
    return StandardNFTGenerator.contract({
      imports,
      contractName: this.name,
      schema: this.schema,
      saveAdminResourceToContractAccount: options?.saveAdminResourceToContractAccount,
    });
  }

  deploy({
    publicKey,
    hashAlgorithm,
    collectionMetadata,
    royalties,
    saveAdminResourceToContractAccount,
  }: {
    publicKey: PublicKey;
    hashAlgorithm: HashAlgorithm;
    collectionMetadata: CollectionMetadata;
    royalties?: Royalty[];
    saveAdminResourceToContractAccount?: boolean;
  }): Transaction<string> {
    return new Transaction(
      ({ imports }: FreshmintConfig) => {
        const script = StandardNFTGenerator.deployToNewAccount({ imports });

        const contractCode = this.getSource(imports, { saveAdminResourceToContractAccount });
        const contractCodeHex = Buffer.from(contractCode, 'utf-8').toString('hex');

        const sigAlgo = publicKey.signatureAlgorithm();

        const { royaltyAddresses, royaltyReceiverPaths, royaltyCuts, royaltyDescriptions } = prepareRoyalties(
          royalties ?? [],
        );

        return {
          script,
          args: [
            fcl.arg(this.name, t.String),
            fcl.arg(contractCodeHex, t.String),
            fcl.arg(publicKey.toHex(), t.String),
            fcl.arg(SignatureAlgorithm.toCadence(sigAlgo), t.UInt8),
            fcl.arg(HashAlgorithm.toCadence(hashAlgorithm), t.UInt8),
            fcl.arg(prepareCollectionMetadata(imports.MetadataViews, collectionMetadata), t.Identity),
            fcl.arg(royaltyAddresses, t.Array(t.Address)),
            fcl.arg(royaltyReceiverPaths, t.Array(t.Path)),
            fcl.arg(royaltyCuts, t.Array(t.UFix64)),
            fcl.arg(royaltyDescriptions, t.Array(t.String)),
            fcl.arg(saveAdminResourceToContractAccount ?? false, t.Bool),
          ],
          computeLimit: 9999,
          signers: this.getSigners(),
        };
      },
      ({ events }: TransactionResult) => {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const accountCreatedEvent = events.find((event) => event.type === 'flow.AccountCreated')!;

        const address = accountCreatedEvent.data['address'];

        this.setAddress(address);

        return address;
      },
    );
  }

  mintNFTs(nfts: MetadataMap[], bucket?: string): Transaction<NFTMintResult[]> {
    return new Transaction(
      ({ imports }: FreshmintConfig) => {
        const script = StandardNFTGenerator.mint({
          imports,
          contractName: this.name,
          contractAddress: this.getAddress(),
          schema: this.schema,
        });

        // Use metadata hash as mint ID
        const mintIds = nfts.map((metadata) => hashMetadata(this.schema, metadata).toString('hex'));

        return {
          script,
          args: [
            fcl.arg(bucket, t.Optional(t.String)),
            fcl.arg(mintIds, t.Array(t.String)),
            ...this.schema.fields.map((field) => {
              return fcl.arg(
                nfts.map((metadata) => field.getValue(metadata)),
                t.Array(field.asCadenceTypeObject()),
              );
            }),
          ],
          computeLimit: 9999,
          signers: this.getSigners(),
        };
      },
      (result) => formatMintResults(result, nfts),
    );
  }
}

function formatMintResults({ events, transactionId }: TransactionResult, nfts: MetadataMap[]): NFTMintResult[] {
  const deposits = events.filter((event) => event.type.includes('.Minted'));

  return deposits.map((deposit, i) => {
    return {
      id: deposit.data.id,
      metadata: nfts[i],
      transactionId,
    };
  });
}
