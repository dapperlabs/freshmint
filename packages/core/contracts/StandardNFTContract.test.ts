import { StandardNFTContract } from './StandardNFTContract';
import { MissingContractAddressError } from './NFTContract';
import { FreshmintClaimSaleContract } from './FreshmintClaimSaleContract';

import {
  client,
  contractHashAlgorithm,
  contractPublicKey,
  ownerAuthorizer,
  getTestSchema,
  getTestNFTs,
  setupEmulator,
  teardownEmulator,
  collectionMetadata,
  createAccount,
} from '../testHelpers';

describe('StandardNFTContract', () => {
  beforeAll(setupEmulator);
  afterAll(teardownEmulator);

  const contract = new StandardNFTContract({
    name: 'StandardNFT_Test',
    schema: getTestSchema(),
    owner: ownerAuthorizer,
  });

  it('should generate a contract', async () => {
    expect(contract.getSource(client.config.imports)).toMatchSnapshot();
  });

  const nfts = getTestNFTs(3);

  it('should fail to mint NFTs before contract is deployed', async () => {
    await expect(async () => await client.send(contract.mintNFTs(nfts))).rejects.toThrow(MissingContractAddressError);
  });

  it('should deploy a contract', async () => {
    await client.send(
      contract.deploy({
        publicKey: contractPublicKey,
        hashAlgorithm: contractHashAlgorithm,
        collectionMetadata,
      }),
    );
  });

  // TODO: refactor this test case into a separate test suite
  it('should deploy a contract and save the admin resource to the contract account', async () => {
    const contractWithAdmin = new StandardNFTContract({
      name: 'StandardNFT_Test_ContractAdmin',
      schema: getTestSchema(),
      owner: ownerAuthorizer,
    });

    await client.send(
      contractWithAdmin.deploy({
        publicKey: contractPublicKey,
        hashAlgorithm: contractHashAlgorithm,
        collectionMetadata,
        saveAdminResourceToContractAccount: true,
      }),
    );
  });

  it('should mint NFTs', async () => {
    await client.send(contract.mintNFTs(nfts));
  });

  const sale = new FreshmintClaimSaleContract(contract);

  const allowlistName = 'default';

  it('should be able to add an address we do not control to the allowlist', async () => {
    // When calling addToAllowlist for the first time with a new name,
    // Freshmint will automatically create the allowlist if it does not exist.
    //
    // This call creates an allowlist with name "default" and adds address 0x0ae53cb6e3f42a79.
    await client.send(
      sale.addToAllowlist({
        name: allowlistName,
        addresses: ['0x0ae53cb6e3f42a79'],
        claims: 1, // this address will be allowed to claim 1 NFT
      }),
    );
  });

  it('should start a sale with an allowlist', async () => {
    await client.send(sale.start({ id: 'default', price: '10.0', allowlist: allowlistName }));
  });

  it('should not be able to claim an NFT when not on the allowlist', async () => {
    await expect(async () => {
      await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, 'default'));
    }).rejects.toThrow();
  });

  it('should be able to add our address to the allowlist', async () => {
    await client.send(
      sale.addToAllowlist({
        name: allowlistName,
        addresses: [ownerAuthorizer.address],
        claims: 2, // our address will be allowed to claim 2 NFTs
      }),
    );
  });

  it('should be able to claim two NFTs when on the allowlist', async () => {
    // Claim the first NFT
    await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, 'default'));

    // Claim the second NFT
    await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, 'default'));
  });

  it('should not be able to claim a third NFT', async () => {
    await expect(async () => {
      await client.send(sale.claimNFT(ownerAuthorizer.address, ownerAuthorizer, 'default'));
    }).rejects.toThrow();
  });

  it('should stop a sale', async () => {
    await client.send(sale.stop('default'));
  });
});
