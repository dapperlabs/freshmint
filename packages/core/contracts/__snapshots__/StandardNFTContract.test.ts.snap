// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardNFTContract should generate a contract 1`] = `
"import NonFungibleToken from 0xf8d6e0586b0a20c7
import MetadataViews from 0xf8d6e0586b0a20c7
import FungibleToken from 0xee82856bf20e2aa6

pub contract StandardNFT_Test: NonFungibleToken {

    pub let version: String

    pub event ContractInitialized()
    pub event Withdraw(id: UInt64, from: Address?)
    pub event Deposit(id: UInt64, to: Address?)
    pub event Minted(id: UInt64)
    pub event Burned(id: UInt64)

    pub let CollectionStoragePath: StoragePath
    pub let CollectionPublicPath: PublicPath
    pub let CollectionPrivatePath: PrivatePath
    pub let AdminStoragePath: StoragePath

    /// The total number of StandardNFT_Test NFTs that have been minted.
    ///
    pub var totalSupply: UInt64

    /// A list of royalty recipients that is attached to all NFTs
    /// minted by this contract.
    ///
    access(contract) let royalties: [MetadataViews.Royalty]
    
    /// Return the royalty recipients for this contract.
    ///
    pub fun getRoyalties(): [MetadataViews.Royalty] {
        return StandardNFT_Test.royalties
    }

    /// The collection-level metadata for all NFTs minted by this contract.
    ///
    pub let collectionMetadata: MetadataViews.NFTCollectionDisplay

    pub struct Metadata {

        pub let name: String
        pub let description: String
        pub let thumbnail: String
        pub let serialNumber: UInt64

        init(
            name: String,
            description: String,
            thumbnail: String,
            serialNumber: UInt64,
        ) {
            self.name = name
            self.description = description
            self.thumbnail = thumbnail
            self.serialNumber = serialNumber
        }
    }

    pub resource NFT: NonFungibleToken.INFT, MetadataViews.Resolver {

        pub let id: UInt64
        pub let metadata: Metadata

        init(metadata: Metadata) {
            self.id = self.uuid
            self.metadata = metadata
        }

        pub fun getViews(): [Type] {
            return [
                Type<MetadataViews.NFTView>(),
                Type<MetadataViews.Display>(),
                Type<MetadataViews.ExternalURL>(),
                Type<MetadataViews.NFTCollectionDisplay>(),
                Type<MetadataViews.NFTCollectionData>(),
                Type<MetadataViews.Royalties>(),
                Type<MetadataViews.Serial>()
            ]
        }

        pub fun resolveView(_ view: Type): AnyStruct? {
            switch view {
                case Type<MetadataViews.NFTView>():
                    return self.resolveNFTView(self.metadata)
                case Type<MetadataViews.Display>():
                    return self.resolveDisplay(self.metadata)
                case Type<MetadataViews.ExternalURL>():
                    return self.resolveExternalURL()
                case Type<MetadataViews.NFTCollectionDisplay>():
                    return self.resolveNFTCollectionDisplay()
                case Type<MetadataViews.NFTCollectionData>():
                    return self.resolveNFTCollectionData()
                case Type<MetadataViews.Royalties>():
                    return self.resolveRoyalties()
                case Type<MetadataViews.Serial>():
                    return self.resolveSerial(self.metadata)
            }

            return nil
        }

        pub fun resolveNFTView(_ metadata: Metadata): MetadataViews.NFTView {
            return MetadataViews.NFTView(
                id: self.id,
                uuid: self.uuid,
                display: self.resolveDisplay(metadata),
                externalURL: self.resolveExternalURL(),
                collectionData: self.resolveNFTCollectionData(),
                collectionDisplay: self.resolveNFTCollectionDisplay(),
                royalties : self.resolveRoyalties(),
                traits: nil
            )
        }
        
        pub fun resolveDisplay(_ metadata: Metadata): MetadataViews.Display {
            return MetadataViews.Display(
                name: metadata.name,
                description: metadata.description,
                thumbnail: MetadataViews.IPFSFile(cid: metadata.thumbnail, path: nil)
            )
        }
        
        pub fun resolveExternalURL(): MetadataViews.ExternalURL {
            let collectionURL = StandardNFT_Test.collectionMetadata.externalURL.url
            let nftOwner = self.owner!.address.toString()
            let nftID = self.id.toString()
            return MetadataViews.ExternalURL(collectionURL.concat(\\"/nfts/\\").concat(nftOwner).concat(\\"/\\").concat(nftID))
        }
        
        pub fun resolveNFTCollectionDisplay(): MetadataViews.NFTCollectionDisplay {
            return StandardNFT_Test.collectionMetadata
        }
        
        pub fun resolveNFTCollectionData(): MetadataViews.NFTCollectionData {
            return MetadataViews.NFTCollectionData(
                storagePath: StandardNFT_Test.CollectionStoragePath,
                publicPath: StandardNFT_Test.CollectionPublicPath,
                providerPath: StandardNFT_Test.CollectionPrivatePath,
                publicCollection: Type<&StandardNFT_Test.Collection{StandardNFT_Test.StandardNFT_TestCollectionPublic}>(),
                publicLinkedType: Type<&StandardNFT_Test.Collection{StandardNFT_Test.StandardNFT_TestCollectionPublic, NonFungibleToken.CollectionPublic, NonFungibleToken.Receiver, MetadataViews.ResolverCollection}>(),
                providerLinkedType: Type<&StandardNFT_Test.Collection{StandardNFT_Test.StandardNFT_TestCollectionPublic, NonFungibleToken.CollectionPublic, NonFungibleToken.Provider, MetadataViews.ResolverCollection}>(),
                createEmptyCollectionFunction: (fun (): @NonFungibleToken.Collection {
                    return <-StandardNFT_Test.createEmptyCollection()
                })
            )
        }
        
        pub fun resolveRoyalties(): MetadataViews.Royalties {
            return MetadataViews.Royalties(StandardNFT_Test.getRoyalties())
        }
        
        pub fun resolveSerial(_ metadata: Metadata): MetadataViews.Serial {
            return MetadataViews.Serial(metadata.serialNumber)
        }
        
        destroy() {
            StandardNFT_Test.totalSupply = StandardNFT_Test.totalSupply - (1 as UInt64)

            emit Burned(id: self.id)
        }
    }

    pub resource interface StandardNFT_TestCollectionPublic {
        pub fun deposit(token: @NonFungibleToken.NFT)
        pub fun getIDs(): [UInt64]
        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT
        pub fun borrowStandardNFT_Test(id: UInt64): &StandardNFT_Test.NFT? {
            post {
                (result == nil) || (result?.id == id):
                    \\"Cannot borrow StandardNFT_Test reference: The ID of the returned reference is incorrect\\"
            }
        }
    }

    pub resource Collection: StandardNFT_TestCollectionPublic, NonFungibleToken.Provider, NonFungibleToken.Receiver, NonFungibleToken.CollectionPublic, MetadataViews.ResolverCollection {
        
        /// A dictionary of all NFTs in this collection indexed by ID.
        ///
        pub var ownedNFTs: @{UInt64: NonFungibleToken.NFT}

        init () {
            self.ownedNFTs <- {}
        }

        /// Remove an NFT from the collection and move it to the caller.
        ///
        pub fun withdraw(withdrawID: UInt64): @NonFungibleToken.NFT {
            let token <- self.ownedNFTs.remove(key: withdrawID) 
                ?? panic(\\"Requested NFT to withdraw does not exist in this collection\\")

            emit Withdraw(id: token.id, from: self.owner?.address)

            return <- token
        }

        /// Deposit an NFT into this collection.
        ///
        pub fun deposit(token: @NonFungibleToken.NFT) {
            let token <- token as! @StandardNFT_Test.NFT

            let id: UInt64 = token.id

            // add the new token to the dictionary which removes the old one
            let oldToken <- self.ownedNFTs[id] <- token

            emit Deposit(id: id, to: self.owner?.address)

            destroy oldToken
        }

        /// Return an array of the NFT IDs in this collection.
        ///
        pub fun getIDs(): [UInt64] {
            return self.ownedNFTs.keys
        }

        /// Return a reference to an NFT in this collection.
        ///
        /// This function panics if the NFT does not exist in this collection.
        ///
        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT {
            return (&self.ownedNFTs[id] as &NonFungibleToken.NFT?)!
        }

        /// Return a reference to an NFT in this collection
        /// typed as StandardNFT_Test.NFT.
        ///
        /// This function returns nil if the NFT does not exist in this collection.
        ///
        pub fun borrowStandardNFT_Test(id: UInt64): &StandardNFT_Test.NFT? {
            if self.ownedNFTs[id] != nil {
                let ref = (&self.ownedNFTs[id] as auth &NonFungibleToken.NFT?)!
                return ref as! &StandardNFT_Test.NFT
            }

            return nil
        }

        /// Return a reference to an NFT in this collection
        /// typed as MetadataViews.Resolver.
        ///
        /// This function panics if the NFT does not exist in this collection.
        ///
        pub fun borrowViewResolver(id: UInt64): &AnyResource{MetadataViews.Resolver} {
            let nft = (&self.ownedNFTs[id] as auth &NonFungibleToken.NFT?)!
            let nftRef = nft as! &StandardNFT_Test.NFT
            return nftRef as &AnyResource{MetadataViews.Resolver}
        }

        destroy() {
            destroy self.ownedNFTs
        }
    }

    /// Return a new empty collection.
    ///
    pub fun createEmptyCollection(): @NonFungibleToken.Collection {
        return <- create Collection()
    }

    /// The administrator resource used to mint and reveal NFTs.
    ///
    pub resource Admin {

        /// Mint a new NFT.
        ///
        /// To mint an NFT, specify a value for each of its metadata fields.
        ///
        pub fun mintNFT(
            name: String,
            description: String,
            thumbnail: String,
            serialNumber: UInt64,
        ): @StandardNFT_Test.NFT {

            let metadata = Metadata(
                name: name,
                description: description,
                thumbnail: thumbnail,
                serialNumber: serialNumber,
            )

            let nft <- create StandardNFT_Test.NFT(metadata: metadata)

            emit Minted(id: nft.id)

            StandardNFT_Test.totalSupply = StandardNFT_Test.totalSupply + (1 as UInt64)

            return <- nft
        }
    }

    /// Return a public path that is scoped to this contract.
    ///
    pub fun getPublicPath(suffix: String): PublicPath {
        return PublicPath(identifier: \\"StandardNFT_Test_\\".concat(suffix))!
    }

    /// Return a private path that is scoped to this contract.
    ///
    pub fun getPrivatePath(suffix: String): PrivatePath {
        return PrivatePath(identifier: \\"StandardNFT_Test_\\".concat(suffix))!
    }

    /// Return a storage path that is scoped to this contract.
    ///
    pub fun getStoragePath(suffix: String): StoragePath {
        return StoragePath(identifier: \\"StandardNFT_Test_\\".concat(suffix))!
    }

    /// Return a collection name with an optional bucket suffix.
    ///
    pub fun makeCollectionName(bucketName maybeBucketName: String?): String {
        if let bucketName = maybeBucketName {
            return \\"Collection_\\".concat(bucketName)
        }

        return \\"Collection\\"
    }

    /// Return a queue name with an optional bucket suffix.
    ///
    pub fun makeQueueName(bucketName maybeBucketName: String?): String {
        if let bucketName = maybeBucketName {
            return \\"Queue_\\".concat(bucketName)
        }

        return \\"Queue\\"
    }

    priv fun initAdmin(admin: AuthAccount) {
        // Create an empty collection and save it to storage
        let collection <- StandardNFT_Test.createEmptyCollection()

        admin.save(<- collection, to: StandardNFT_Test.CollectionStoragePath)

        admin.link<&StandardNFT_Test.Collection>(StandardNFT_Test.CollectionPrivatePath, target: StandardNFT_Test.CollectionStoragePath)

        admin.link<&StandardNFT_Test.Collection{NonFungibleToken.CollectionPublic, StandardNFT_Test.StandardNFT_TestCollectionPublic, MetadataViews.ResolverCollection}>(StandardNFT_Test.CollectionPublicPath, target: StandardNFT_Test.CollectionStoragePath)
        
        // Create an admin resource and save it to storage
        let adminResource <- create Admin()

        admin.save(<- adminResource, to: self.AdminStoragePath)
    }

    init(collectionMetadata: MetadataViews.NFTCollectionDisplay, royalties: [MetadataViews.Royalty], admin: AuthAccount) {

        self.version = \\"0.2.3\\"

        self.CollectionPublicPath = StandardNFT_Test.getPublicPath(suffix: \\"Collection\\")
        self.CollectionStoragePath = StandardNFT_Test.getStoragePath(suffix: \\"Collection\\")
        self.CollectionPrivatePath = StandardNFT_Test.getPrivatePath(suffix: \\"Collection\\")

        self.AdminStoragePath = StandardNFT_Test.getStoragePath(suffix: \\"Admin\\")

        self.royalties = royalties
        self.collectionMetadata = collectionMetadata

        self.totalSupply = 0

        self.initAdmin(admin: admin)

        emit ContractInitialized()
    }
}
"
`;
